sdk/src/main/scala/com/horizen/api/http/SidechainWalletApiRoute.scala

```diff
@@ -134,6 +138,122 @@ case class SidechainWalletApiRoute(override val settings: RESTApiSettings,
+          val optionalPrivKey = wallet.secretByPublicKey(proposition)
+          if (optionalPrivKey.isEmpty) {
+            ApiResponseUtil.toResponse(ErrorPropositionNotFound("Proposition not found in the wallet!", JOptional.empty()))
+          } else {
+            ApiResponseUtil.toResponse(RespExportSecret(BytesUtils.toHexString(sidechainSecretsCompanion.toBytes(optionalPrivKey.get()))))
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Perform a dump on a file of all the secrets inside the wallet.
+   */
+  def dumpSecrets: Route = (post & path("dumpSecrets")) {
+    withAuth {
+      entity(as[ReqDumpWallet]) { body =>
+        val writer = new PrintWriter(new File(body.path))
+        writer.write(s"# Secrets dump created on ${java.time.Instant.now()} \n")
+        withNodeView { sidechainNodeView =>
+          val wallet = sidechainNodeView.getNodeWallet
+          wallet.allSecrets().forEach(key =>
+            writer.write(BytesUtils.toHexString(sidechainSecretsCompanion.toBytes(key))+" "+BytesUtils.toHexString(key.publicImage().bytes())+"\n")
+          )
+          writer.close()
+          ApiResponseUtil.toResponse(RespDumpSecrets(s"Secrets dump completed successfully at: ${body.path}"))
+        }
+      }
+    }
+  }
+
+  /**
+   * Import all secrets contained in a file.
+   * The file format should be equal to the file format generated by the endpoint dumpSecrets. (SECRETS + " " + PUBLICKEY)
+   */
+  def importSecrets: Route = (post & path("importSecrets")) {
+    withAuth {
+      entity(as[ReqDumpWallet]) { body =>
+        val reader = new Scanner(new File(body.path))
+
+        //First collect every secrets and verify that their public image match with the corresponding public key in the file.
+        var lineNumber = 1
+        val secrets = new JArrayList[(SidechainTypes#SCS, Int)]()
+        var error: JOptional[ErrorResponse] = JOptional.empty()
+        while (reader.hasNextLine && error.isEmpty) {
+          val line = reader.nextLine()
+          if (!line.contains("#")) {
+            val keyPair = line.split(" ")
+            sidechainSecretsCompanion.parseBytesTry(BytesUtils.fromHexString(keyPair(0))) match {
+              case Success(value) =>
+                if(!BytesUtils.toHexString(value.publicImage().bytes()).equals(keyPair(1))) {
+                  log.error(s"Import Wallet: Public key doesn't match: ${BytesUtils.toHexString(value.publicImage().bytes())}  ${keyPair(1)}")
+                  error = JOptional.of(ErrorPropositionNotMatch(s"Public key doesn't match on line ${lineNumber}", JOptional.empty()))
+                } else {
+                  secrets.add((value, lineNumber))
+                }
+              case Failure(e) =>
+                log.error(s"Import Wallet: Failed to parse the secret: ${keyPair(0)}")
+                error = JOptional.of(ErrorFailedToParseSecret(s"Failed to parse the secret at line ${lineNumber}", JOptional.of(e)))
```

Comment:
in case we have multiple errors, we track only the last one.
Let's track the whole list, or stop parsing after the first error.

Reply:
I missed that point.
Honestly, "error" processing flow can be simplified just by `return ApiResponseUtil.toResponse(error)` here.
Currently there is one place where we declare optional error field, another where we check if it is not empty in the while loop, another where we setup, and one more where we check and return error result.